### PR TITLE
fix: exclude duplicate calls for class hash and entry points

### LINKED ISSUE #85: Exclude Same sub-calls in Different parent calls (Mandatory)

### PR DESCRIPTION
**Pull request by marcuspang**
Fixes #85

## Changes

1. `get_calls_with_count` now returns a `HashMap<(EntryPoint, ClassHash), usize>`
    - this was previously called `get_calls`, which returned a list of possible tuple values, but I chose to return the concrete `(EntryPoint, ClassHash)` pairs instead for clarity
2. updated call `Value` parsing, which can now parse keys that have `"Same(0x123)"` values
    - previously, the function could only parse from `"0x123"` values, which meant that the parsing of some calls terminated early because of the `"Same"` prefix 
3. update unit tests for `get_calls_with_count` and added some for helper functions

## Notes

- Currently, there is an edge case when there are nested calls in different depths with the same entry point and class hash tuple, which will be resolved in #91
- If base has `[(ch1, ep1), (ch2, ep2)]` and native has `[(ch1, ep1)]`, we only count `(ch1, ep1)` pair once
- When native has no calls, i.e.  `"Empty"` or `[]`, then `get_calls_with_count` will return `[]`

### CODE REVIEW

File: src/gather_classes.rs [hunk 1]
```diff
@@ -257,80 +259,148 @@ fn update_report(
...
+            .entry(key)
+            .and_modify(|existing_count| *existing_count += count)
+            .or_insert(count);
+    }
+
+    merged_map.into_iter().collect()
+}
+
+/// Converts recursive calls to value.get("calls") to a boxed iterator.
+fn get_calls_with_count(obj: &Value) -> Result<Vec<CallWithCount>, anyhow::Error> {
```
Comment (marcuspang) [inline src/gather_classes.rs:292]: Will change back to Iter once logic is correct

**Comment by rodrigo-pino** [comment]
> How should conflicting calls be handled? If base has [(ch1, ep1), (ch2, ep2)] and native has [(ch1, ep1)], do we return a count of 1 or 2 for (ch1, ep1) (currently taking the sum in merge_calls_with_count)

A count of 1 for (ch1, ep1). We ignore (ch2, ep2) since there is no data

**Comment by rodrigo-pino** [comment]
If (ch1, ep1) and (ch1, ep2) that's a different Pair and cannot be compared. You can only compare "same keys" like when (ch_i, ep_k) exists in both.

File: src/gather_classes.rs (hunk 1 above)
Comment (rodrigo-pino) [reply src/gather_classes.rs:292]: You don't have to keep iter. In my opinion is a bit of an unnecessary abstraction, you can return what you think is best.

File: src/gather_classes.rs [hunk 2]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
     assert!(report.blocks.insert(block_num));
     Ok(updates)
 }
 
-// Converts recursive calls to value.get("calls") to a boxed iterator.
-fn get_calls<'a>(obj: &'a Value) -> Result<Box<dyn Iterator<Item = &Value> + 'a>, anyhow::Error> {
+/// Merges two lists of calls into a single list of calls. We only consider a [CallWithCount] if
+/// it exists in both lists.
+///
+/// This means that if native is empty, then no calls will be added to the merged list.
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:272]:
```suggestion
/// For example if either base or native is empty, then no calls from the other will be added to the merged list
```

File: src/gather_classes.rs [hunk 3]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
+    for (call_key, base_count) in base_calls {
+        if native_keys.contains(&call_key) {
+            merged_map
+                .entry(call_key)
+                .and_modify(|count| *count += base_count)
+                .or_insert(base_count);
+        }
+    }
+
+    merged_map.into_iter().collect()
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:292]: Why are you using a vector of tuples, instead of a HashMap?

Signaling it in this line, but it applies to the whole solution. Why don't just deal with a "Map" directly?

File: src/gather_classes.rs [hunk 4]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
+                    get_calls_inner(call, result)?;
+                }
+                Ok(())
+            }
+            Value::Object(obj_map) => {
+                // If call is different, then base and native will have their own list of calls
+                if let Some(Value::Object(diff_map)) = obj_map.get(DIFFERENT) {
+                    let base_list = diff_map
+                        .get("base")
+                        .ok_or(anyhow!("difference should have base"))?;
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:322]: _Different_ should have [base|native]

File: src/gather_classes.rs [hunk 5]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
+                    let mut native_calls = Vec::new();
+                    get_calls_inner(native_list, &mut native_calls)?;
+
+                    result.extend(merge_calls_with_count(base_calls, native_calls));
+                } else {
+                    match get_call_with_count(obj) {
+                        Ok(current_call) => {
+                            result.push(current_call);
+                        }
+                        // We let the parsing fail silently, as there may be inner calls that are valid
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:338]: Could you expand on this explanation?

File: src/gather_classes.rs [hunk 6]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
-            .get(key)
-            .context(format!("call: `{call}`\n-------Extracting key {key}"))?;
-        let felt_version = serde_json::from_value(next.clone())
-            .context(format!("convert value `{}` to felt", next))?;
-        values.push(felt_version);
+/// Retrieves a [CallWithCount] from a call object.
+///
+/// If any of the keys are Different, then the call is considered invalid.
+fn get_call_with_count(call: &Value) -> Result<CallWithCount, anyhow::Error> {
+    let keys = ["entry_point_selector", "class_hash"];
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:363]: Why not use `const`?

File: src/gather_classes.rs [hunk 7]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
+            let felt_version = FieldElement::from_hex_be(value)
+                .context(format!("Failed to convert value to felt"))?;
+            values.push(felt_version);
+        } else {
+            return Err(anyhow!("Failed to parse call {call} for key: {key}"));
+        }
     }
-    Ok((values[0], values[1]))
+
+    Ok(((values[0], values[1]), 1))
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:382]: This is a bit hard to read. Let's improve readability of the code so that by just reading one can understand what's being returned there.

**Review by rodrigo-pino — COMMENTED** [review]
First round of review. Will give it another look once current comments are solved.

Solution looks good, it solves the problem. We can make it more polished and apply some programming concepts to make it more readable.

**Review by PearsonWhite — CHANGES_REQUESTED** [review]

File: src/gather_classes.rs (hunk 3 above)
Comment (marcuspang) [reply src/gather_classes.rs:292]: Initially, I kept it as a vector because it was easier to combine instead of creating multiple hashmaps. But now that I'm not chaining vecs, I think it's better to use HashMaps

File: src/gather_classes.rs (hunk 6 above)
Comment (marcuspang) [reply src/gather_classes.rs]: I didn't use it initially because I had to specify the length of the `keys` array in the type, but I think I will choose not to create an array instead. Adding the array doesn't have any benefits

File: src/gather_classes.rs (hunk 7 above)
Comment (marcuspang) [reply src/gather_classes.rs]: I made the method more functional, I think it's clearer to understand at first glance now

File: src/gather_classes.rs [hunk 8]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
+            updates += count;
         }
     }
     assert!(report.blocks.insert(block_num));
     Ok(updates)
 }
 
-// Converts recursive calls to value.get("calls") to a boxed iterator.
-fn get_calls<'a>(obj: &'a Value) -> Result<Box<dyn Iterator<Item = &Value> + 'a>, anyhow::Error> {
+/// Merges two lists of calls into a single list of calls. We only consider a [CallWithCount] if
```
[the remark this answers is not in the document]
Comment (marcuspang) [reply src/gather_classes.rs]: I'm not sure I understand, some `CallKey`s may appear twice, e.g. when a call with `CallKey` has some inner call with the same `CallKey`. Currently, I have chosen to refactor it to use hash maps, so I hope it is clearer now

File: src/gather_classes.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcuspang) [reply src/gather_classes.rs]: Thank you, this is a good example

File: src/gather_classes.rs [hunk 9]
```diff
@@ -257,80 +255,131 @@ fn update_report(
...
+                    let native_list = diff_map
+                        .get("native")
+                        .ok_or(anyhow!("difference should have native"))?;
+
+                    let mut base_calls = Vec::new();
+                    get_calls_inner(base_list, &mut base_calls)?;
+                    let mut native_calls = Vec::new();
+                    get_calls_inner(native_list, &mut native_calls)?;
+
+                    result.extend(merge_calls_with_count(base_calls, native_calls));
```
[the remark this answers is not in the document]
Comment (marcuspang) [reply src/gather_classes.rs]: This implementation doesn't really consider the different layers for the calls. Right now this would naively just compare 2 dictionaries and return `{ call_1: 1, call_3: 1 }`.

Do you think call 2 and call 3 should be removed entirely since they are of different depths?

File: src/gather_classes.rs [hunk 10]
```diff
@@ -257,80 +254,127 @@ fn update_report(
...
     // TODO (#72) Put debug logging here under a core::option_env flag so it is normally hidden.
-    match obj {
-        Value::String(string) => {
-            if string_is_same(string) {
-                Ok(Box::new(::std::iter::empty()))
-            } else {
-                Err(anyhow!("unexpected string: {}", string))
+    fn get_calls_inner(
+        obj: &Value,
+        mut result: &mut HashMap<CallKey, usize>,
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:295]: Suggestion: Apply FP programming principles

File: src/gather_classes.rs [hunk 11]
```diff
@@ -257,80 +254,127 @@ fn update_report(
...
+///
+/// If any of the keys are Different, then the call is considered invalid and this function will return an Error.
+fn get_call_key(call: &Value) -> Result<CallKey, anyhow::Error> {
+    let parse_field = |key: &str| -> Result<FieldElement, anyhow::Error> {
+        call.get(key)
+            .and_then(Value::as_str)
+            .ok_or_else(|| anyhow!("Failed to parse call {call} for key: {key}"))
+            .and_then(|s| {
+                if string_is_same(s) {
+                    parse_same_string(s).context(anyhow!("Failed to parse SAME value"))
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:364]: If string_is_same parse same should be correct by default, right.

If wouldn't ask, I would directly perform a match:

```rust
match parse_same_string(s) {
   Ok() => "SAME"
   Not Ok() => s
}
```

File: src/gather_classes.rs [hunk 12]
```diff
@@ -257,80 +254,127 @@ fn update_report(
...
-            .context(format!("call: `{call}`\n-------Extracting key {key}"))?;
-        let felt_version = serde_json::from_value(next.clone())
-            .context(format!("convert value `{}` to felt", next))?;
-        values.push(felt_version);
-    }
-    Ok((values[0], values[1]))
+/// Retrieves a [CallKey] from a call object.
+///
+/// If any of the keys are Different, then the call is considered invalid and this function will return an Error.
+fn get_call_key(call: &Value) -> Result<CallKey, anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (marcuspang) [reply src/gather_classes.rs:357]: Yes, it's a reference here

File: src/gather_classes.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcuspang) [reply src/gather_classes.rs]: I will leave it as another ticket to resolve #91

File: src/gather_classes.rs (hunk 5 above)
Comment (marcuspang) [reply src/gather_classes.rs]: Updated the comment with Pearson's description

File: src/gather_classes.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (rodrigo-pino) [reply src/gather_classes.rs:355]: Why don't take ownership then?

File: src/gather_classes.rs [hunk 13]
```diff
@@ -257,80 +254,123 @@ fn update_report(
...
+        mut result: HashMap<CallKey, usize>,
+    ) -> Result<HashMap<CallKey, usize>, anyhow::Error> {
+        match obj {
+            Value::String(string) => {
+                // If the string is SAME or EMPTY, then there are no calls to extract and continue processing
+                if string_is_same(string) || string_is_empty(string) {
+                    Ok(result)
+                } else {
+                    Err(anyhow!("String is not SAME or EMPTY: {}", string))
+                }
```
Comment (rodrigo-pino) [inline src/gather_classes.rs:305]: Sorry didn't notice before, but this case, after the previous change, becomes redundant. 

All none matching gases are handled in the branch `_ => { .. } `

**Review by rodrigo-pino — APPROVED** [review]
Looks good. Address comments and merge at will

File: src/gather_classes.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (marcuspang) [reply src/gather_classes.rs:355]: Converting this was a bit tricky for me, I tried to take in both as `Value` and `Map<String, Value>`, but that requires ownership of the input `call` parameter in `get_calls_with_count`. I couldn't get it to work nicely without doing some form of `.clone()`. 

I updated it to take `Map<String, Value>` instead
